### ISSUE TITLE
Adaption for mobile linux platforms.

### ISSUE THREAD

**Issue by Alexmitter** [flathub/com.github.jms55.Sandbox#2]
Hello

I am part of the Mobian Project, we build a debian based linux distribution that has a phone shell.
It would be nice to run this fun little game on it as I remember a similar game many years ago from my old Windows Mobile PDA.
Two things would need to be done to make this a working.

1. A aarch64 build on flathub
2. Touchable buttons to select the various Particle Types and functions.
3. The frame would need to adapt to the form phone screen form factor.

**Comment by JMS55** [comment]
Hi, I'd love for this to run on mobile. Thank you for taking an interest.

1. The SIMD noise based library I use dosen't support aarch64 SIMD as far as I can tell, so I hope the fallback is fast enough to work.
2. This is coming, however I'm not sure when. It's currently my top priority, along with in-game instructions, once I finish investigating a slowdown bug.
3. I believe I just need to start the game in fullscreen on mobile devices. However, what are the resolutions phones typically use? The game is 480x270 at it's smallest, and snaps to integer multiples of that size. If phones aren't a multiple of this, there will be black bars around it.
4. Additionally, Mobian uses the Phosh shell from what I can see, which uses Wayland right? The winit library I use supports Wayland, but I force X11, as winit currently has bugs around decorations, full screen with gnome shell, and keypresses on Wayland. This would have to be changed, but it should be safe to change for mobile.
5. I'm more concerned about wgpu, the rendering library I use. Do phones support Vulkan? OpenGL is currently not an option with wgpu.
6. I don't currently handle touch events, only cursor events. I'd have to fix this.

**Comment by Alexmitter** [comment]
1. This would be a bummer but I could try find some help for that.
2. No issue with not putting this on your prio list at all.
3. Current development target are the Pinephone and the Librem 5, both have a ‎720×1440 screen with a scaling factor of 2, so the resolution divided by two. Phosh' compositor has a scale-to-fit function that could be used as a workaround if the window cant snap into this resolution.
4. Yes, phosh uses Wayland, Xwayland is also available but I would not recommend to target it at all. But no window decoration beside what the wayland clients them self draw is displayed.
5. This is a major breaking point as there is currently no Vulkan support in the driver we are using, This will take more time, a lot more time. If your grapics library does not support OpenGL, there is about no chance to wire phone support up now.

Looks like there are too many breaking points currently, so I would recommend to keep this in mind and when the time is right and your graphics library gains support, it would make sense to continue.

**Comment by JMS55** [comment]
Yes, while most of these issues are fixable, OpenGL support seems like it's still going to take some time[0]. I'll keep mobile in mind, and hopefully later it will be possible!

0: https://github.com/gfx-rs/wgpu/issues/450

**Comment by JMS55** [comment]
Update:
* The UI is basically done, minus a very small bug
* Wayland support will be added very soon
* Wgpu is getting OpenGL support soon-ish (assuming no huge bugs popup)

I plan to make a new release once I fix the UI bug, and then winit and freedesktop-sdk make a new release.
After that, once wgpu gets OpenGL support, I'll add support for touch controls and hopefully that's all that's needed for an aarch64/mobile build.

**Comment by Alexmitter** [comment]
It looks like its not going to happen on the Wgpu side, the Pinephone would support either OpenGL 2.1 or GLES 2.0 and it looks like support for both is not considered, only GLES 3.0 maybe.

The Wayland support on the other side is nice. Just one tip for that, it may be worth looking at libdecoration to gain native looking window decoration :)

**Comment by JMS55** [comment]
Yeah, unfortunately it turns out I was optimistic about OpenGL support.

I've looked at libdecoration before, but there's little documentation and no rust bindings. I might come back to it in the future, but not at the moment. winit is also updating sctk eventually, the wayland crate it uses, which is going to change how decorations look. I'm planning to reevaluate this after that lands.

Closing this issue as it doesn't look like it's going to be feasible for mobile support, but eventually web support should land when webgpu the spec is stabilized, which may or may not be fast enough to be usable from a phone.
